Add support for renaming Distributed Tables

DimCitus opened the pull request:
This Pull Request implements support for renaming Distributed Tables, fixing #1249. It also adds a commit to edit the regression tests and have more coverage of non-default schema (not public, not in the search_path).

It might be better for me to rebase/squash the main commit and its follow-up make reindent pass though, but will that trigger another travis-ci run?

marcocitus commented on src/backend/distributed/relay/relay_event_utility.c:
```diff
@@ -644,7 +667,6 @@ AppendShardIdToName(char **name, uint64 shardId)
 	{
 		snprintf(extendedName, NAMEDATALEN, "%s%s", (*name), shardIdAndSeparator);
 	}
-
```
this is the line `citus_indent` is complaining about

marcocitus commented on src/test/regress/expected/multi_reference_table.out:
```diff
@@ -1308,26 +1308,26 @@ SELECT logicalrelid FROM pg_dist_shard WHERE logicalrelid::regclass::text LIKE '
-CREATE INDEX reference_index_1 ON reference_table_ddl(value_1);
-CREATE INDEX reference_index_2 ON reference_table_ddl(value_2, value_3);
+CREATE INDEX reference_index_1 ON reference_schema.reference_table_ddl(value_1);
+CREATE INDEX reference_index_2 ON reference_schema.reference_table_ddl(value_2, value_3);
```
couldn't we just use search_path here? that's the pattern followed by some other tests.

DimCitus replied:
Depends on what exactly we want to test. Is the test more interesting with using the search_path or without using it?

marcocitus replied:
Both are interesting. Anyway, no strong objections to keep it this way.

marcocitus commented on src/backend/distributed/relay/relay_event_utility.c:
```diff
@@ -394,6 +394,29 @@ RelayEventExtendNames(Node *parseTree, char *schemaName, uint64 shardId)
+					ereport(ERROR,
+							(errcode(ERRCODE_NAME_TOO_LONG),
+							 errmsg(
+								 "distributed identifier must be less than %d characters",
```
maybe: distributed identifier -> shard name

marcocitus replied:
could maybe also say "shard name %s exceeds %d characters"

marcocitus commented on src/backend/distributed/relay/relay_event_utility.c:
```diff
@@ -394,6 +394,29 @@ RelayEventExtendNames(Node *parseTree, char *schemaName, uint64 shardId)
+				 *
+				 * See also https://github.com/citusdata/citus/issues/1664
+				 */
+				if (strlen(*newRelationName) >= (NAMEDATALEN - 1))
```
A minor Citus convention (which I think derives from Code Complete 2) is to avoid function calls in expressions where possible. So the preferred pattern is:

```
newRelationNameLength = strlen(*newRelationName);
if (newRelationNameLength >= (NAMEDATALEN - 1))
{
   ...
}
```

That way we could set a break point on the `if` and see the actual length that causes the expression to evaluate to true.

marcocitus commented:
would be good to also add a test on distributed tables, as those are a bit different from reference tables.

DimCitus commented:
I added a commit on top of the previous two, in order to make it easier to follow-up in this PR. Let me know if/when it should be squashed into the first commit of this branch.

About distributed tables, we already test them in https://github.com/citusdata/citus/blob/feature/rename-table/src/test/regress/sql/multi_name_lengths.sql#L31 ; see https://github.com/citusdata/citus/blob/feature/rename-table/src/test/regress/sql/multi_name_lengths.sql#L66 ; so I think we're good and didn't add another test in the suite. Is that ok?

Other suggestions applied. `make reindent` still doesn't do much locally:

```
$ uncrustify --version
Uncrustify-0.66.1_f

$ make reindent
cd /Users/dim/dev/CitusData/citus && citus_indent --quiet
/usr/local/etc/citustools/citus-style.cfg:788 Unknown symbol 'align_number_left'
```
```

marcocitus commented:
> ; so I think we're good and didn't add another test in the suite. Is that ok?

this appears to only test a failure case, we have yet to confirm in the regression tests that we can successfully rename a distributed table

marcocitus commented:
@jasonmp85 any thoughts on the uncrustify error? should we update our config?

DimCitus commented:
I just added some test cases in their own patch, and rebased against origin/master that changed in between, so that I could then push force here. Once review is all clear, I'll need to rebase/squash my 4 commits into 2 of them, if I understand correctly. I'm thinking doing that as a last step so that we can see the incremental revisions in the meantime.

marcocitus commented on src/test/regress/input/multi_alter_table_statements.source:
```diff
@@ -356,10 +355,30 @@ FROM
+
+-- show rename worked on one worker, too
+\c - - - :worker_1_port
+SELECT relname FROM pg_class WHERE relname LIKE 'lineitem_alter%';
```
It might make sense to add `ORDER BY relname` to prevent the output from changing in different environments (e.g. Travis). I think we need to change existing cases that omit the ORDER BY, but it's a good practice to follow going forward.

DimCitus replied:
And Travis is failing right on that here. I'm only going to edit new tests in this PR, though.

DimCitus commented:
Oops, in a previous version of the “Tests results should have a known ordering.” I included unwanted changes. Cleaned the patch and git pushed force. Hope that's it now ;-)

marcocitus commented:
Yep, looks good. Squash and :ship:
